Plugins/Data/NWScript/inc_array.nss

```diff
@@ -0,0 +1,485 @@
+void Array_Insert_Obj(string tag, int index, object element, object obj=OBJECT_INVALID);
+void Array_Insert_Str(string tag, int index, string element, object obj=OBJECT_INVALID);
+/// @}
+
+/// @defgroup data_array_pushback Array Pushback
+/// @brief Pushes an element to the back of the collection.
+/// @remark Functionally identical to an insert at index size-1.
+/// @ingroup data
+/// @param obj The object.
+/// @param tag The tag.
+/// @param element The element.
+/// @{
+void Array_PushBack_Flt(string tag, float  element, object obj=OBJECT_INVALID);
+void Array_PushBack_Int(string tag, int    element, object obj=OBJECT_INVALID);
+void Array_PushBack_Obj(string tag, object element, object obj=OBJECT_INVALID);
+void Array_PushBack_Str(string tag, string element, object obj=OBJECT_INVALID);
+/// @}
+
+/// Resizes the array. If the array is shrinking, it chops off elements at the ned.
+void Array_Resize(string tag, int size, object obj=OBJECT_INVALID);
+
+/// Reorders the array such each possible permutation of elements has equal probability of appearance.
+void Array_Shuffle(string tag, object obj=OBJECT_INVALID);
+
+/// Returns the size of the array.
+int Array_Size(string tag, object obj=OBJECT_INVALID);
+
+/// Sorts the collection based on descending order.
+void Array_SortAscending(string tag, int type=TYPE_STRING, object obj=OBJECT_INVALID);
+
+/// Sorts the collection based on descending order.
+void Array_SortDescending(string tag, int type=TYPE_STRING, object obj=OBJECT_INVALID);
+
+/// @defgroup data_array_set Array Set
+/// @brief Sets the element at the index, where size > index >= 0.
+/// @ingroup data
+/// @param obj The object.
+/// @param tag The tag.
+/// @param index The index.
+/// @param element The element.
+/// @{
+void Array_Set_Flt(string tag, int index, float  element, object obj=OBJECT_INVALID);
+void Array_Set_Int(string tag, int index, int    element, object obj=OBJECT_INVALID);
+void Array_Set_Obj(string tag, int index, object element, object obj=OBJECT_INVALID);
+void Array_Set_Str(string tag, int index, string element, object obj=OBJECT_INVALID);
+/// @}
+
+/// @}
+
+//
+// Local Utility Functions.
+//
+string GetTableName(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    return "array_" + ObjectToString(obj) + "_" + tag;
+}
+
+string GetTableCreateString(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    // for simplicy sake, everything is turned into a string.  Possible enhancement
+    // to create specific tables for int/float/whatever.
+    return "create table if not exists " + GetTableName(tag, obj) + " ( ind integer primary key, value text)";
+}
+
+int TableExists(string tag, object obj=OBJECT_INVALID) {
+    if (obj == OBJECT_INVALID)
+        obj = GetModule();
+    string stmt = "SELECT NAME FROM SQLITE_MASTER WHERE TYPE = 'table' AND NAME = '" + GetTableName(tag, obj) + "'";
```

Comment:
When I suggested you capitalize SQL keywords that would not include table and column names, so this line for example would look like:
`string stmt = "SELECT name FROM sqlite_master WHERE type = 'table' AND name = '" + GetTableName(tag, obj) + "'";`
It wasn't a big deal but it just helps with readability